segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorage.java

```diff
@@ -0,0 +1,203 @@
+            this.sealed = true;
+        } else {
+            this.sealed = false;
+        }
+        this.containerEpoch = containerEpoch;
+        this.updateVersion = updateVersion;
+    }
+
+    /**
+     * Returns the BK ledger which has the given offset and is writable.
+     * @param offset offset from which writes start.
+     * @return The metadata of the ledger.
+     */
+    LedgerData getLedgerDataForWriteAt(long offset) throws BadOffsetException {
+        if (offset != length) {
+            throw new BadOffsetException(this.getName(), length, offset);
+        }
+        LedgerData ledgerData = this.getLastLedgerData();
+        if (ledgerData != null && !ledgerData.getLedgerHandle().isClosed()) {
+            return ledgerData;
+        } else {
+            // If there is no ledger, create a new one.
+            LedgerData data = manager.createLedgerAt(this.name, (int) offset);
+            this.dataMap.put((int) offset, data);
+            return data;
+        }
+    }
+
+    /**
+     * Add a new BK ledger and metadata at a given offset.
+     * @param offset The starting offset represented by the ledger.
+     * @param ledgerData metadata of the ledger.
+     */
+    synchronized void addToList(int offset, LedgerData ledgerData) {
+
+        // If we are replacing an existing ledger, adjust the length
+
+        LedgerData older = this.dataMap.put(offset, ledgerData);
+        if (older != null) {
+            this.length -= older.getLedgerHandle().getLength();
+        }
+        this.length += ledgerData.getLedgerHandle().getLength();
+    }
+
+    /**
+     * Increase length of the LogStorage as a side effect of the write operation.
+     * This is just a cache operation. The length is not persisted.
+     *
+     * @param size size of data written.
+     */
+    synchronized void increaseLengthBy(int size) {
+        this.length += size;
+    }
+
+    synchronized CompletableFuture<Void> deleteAllLedgers() {
+        return CompletableFuture.allOf(
+                this.dataMap.entrySet().stream().map(entry -> manager.deleteLedger(entry.getValue().getLedgerHandle())).toArray(CompletableFuture[]::new));
+    }
+
+    LedgerData getLastLedgerData() {
+        if (this.dataMap.isEmpty()) {
+            return null;
+        } else {
+            return this.dataMap.entrySet().stream().max(Comparator.comparingInt(Map.Entry::getKey)).get().getValue();
+        }
+    }
+
+    synchronized void markSealed() {
+        this.sealed = true;
+    }
+
+    /**
+     * Creates a list of curator transaction for merging source LogStorage in to this.
+     * @param source Name of the source ledger.
+     * @return list of curator operations.
+     */
+    synchronized List<CuratorOp> addLedgerDataFrom(LogStorage source) {
+        List<CuratorOp> retVal = source.dataMap.entrySet().stream().map(entry -> {
+            int newKey = (int) (entry.getKey() + this.length);
+            this.dataMap.put(newKey, entry.getValue());
```

Comment:
I'm not sure this will do what you want. This is streaming the whole of the map and adds a second element into the map for each existing element. However it is not guaranteed that stream will or will not see items added to the map after it is started. As such there may be three or four entries added with the same value. (In fact, this is not even guaranteed to not be an infinite loop.)

Reply:
This is adding entries of the source ledger map to the current ledger map with new keys.